2017 Robot/src/org/usfirst/frc/team4999/lights/Animator.java

```diff
@@ -6,106 +6,110 @@
 
 import edu.wpi.first.wpilibj.Timer;
 
+/**
+ * Runs in an infinite loop. Displays a frame of {@link Animation}, then waits the duration specified by the animation
+ * @author jordan
+ *
+ */
 class AnimatorThread extends Thread {
-	private Color[] currentState;
 	private Display out;
 	private Animation current;
-	private double brightness;
 	
-	private final int TIME_TO_SEND_FRAME = 50;
+	// Generic object. The wait() method is used to pause the run thread until the notifyAll() method is called
+	// If an animation returns -1 for the time to wait, the run thread pauses execution until a new animation is set
+	private final Object pauseLock = new Object();
 	
+	private long timeToSend = 50;
 	
-	public AnimatorThread(Display out, Color[] currentState, Animation current, double brightness) {
-		if(brightness < 0 || brightness > 1) throw new IllegalArgumentException("Brightness must be within [0,1]"); 
-		this.currentState = currentState;
+	
+	public AnimatorThread(Display out, Animation current) {
 		this.out = out;
 		this.current = current;
-		this.brightness = brightness;
 		
+		// Needs to be daemon so that it doesn't block the exit of the JVM
 		this.setDaemon(true);
 	}
 	
-	private Color dim(Color in) {
-		return new Color((int)(in.getRed() * brightness), (int)(in.getGreen() * brightness), (int)(in.getBlue() * brightness));
-	}
-	
-	private Color[] setBrightness(Color[] in) {
-		Color[] out = in.clone();
-		for(int i = 0; i < out.length; i++) {
-			out[i] = dim(out[i]);
+	public void setAnimation(Animation newAnimation) {
+		synchronized (current) {
+			this.current = newAnimation;
+		}
+		// Tells the thread to continue. 
+		synchronized (pauseLock) {
+			pauseLock.notifyAll();
 		}
-		return out;
 	}
 	
 	public void run() {
-		while(!Thread.interrupted()) {
-			currentState = current.animate(currentState);
-			out.show(setBrightness(currentState));
-			int delay = current.getDelayUntilNextFrame();
-			//System.out.println("Expected: " + delay);
-			delay -= TIME_TO_SEND_FRAME;
-			delay = (delay < 0) ? 0 : delay;
-			if (delay > 0) Timer.delay(delay / 1000.0);
+		while(!Thread.interrupted()){
+			
+			int delay;
+			synchronized (current) {
+				// show current frame. keep track of how long it took
+				timeToSend = out.show(current.animate());
+				if(timeToSend < 0) {
+					System.out.println("Failed to write to neopixels, suspending animation thread");
```

Comment:
Yeah, I should definitely be calculating the time to send in the Animator thread. I was returning the value instead because I had been calculating the time to send as part of debugging the I2C code, and I hadn't thought I could just calculate the time in the Animator thread. I'll remove the return value, and have the try/catch block just log the error to Sys.err if there is an error.